src/scales/scale.time.js

```diff
@@ -404,21 +404,39 @@ function computeOffsets(table, ticks, min, max, options) {
 	return {start: start, end: end};
 }
 
+function setMajorTicks(scale, indices, ticks, majorUnit) {
+	var adapter = scale._adapter;
+	var first = +adapter.startOf(ticks[0].value, majorUnit);
+	var last = ticks[ticks.length - 1].value;
+	var major, index;
+
+	for (major = first; major <= last; major = +adapter.add(major, 1, majorUnit)) {
+		index = indices[major];
+		if (typeof index !== 'undefined') {
+			ticks[index].major = true;
+		}
+	}
+	return ticks;
+}
+
 function ticksFromTimestamps(scale, values, majorUnit) {
 	var ticks = [];
-	var i, ilen, value, major;
+	var ts = {};
+	var ilen = values.length;
+	var i, value;
 
-	for (i = 0, ilen = values.length; i < ilen; ++i) {
+	for (i = 0; i < ilen; ++i) {
 		value = values[i];
-		major = majorUnit ? value === +scale._adapter.startOf(value, majorUnit) : false;
+		ts[value] = i;
 
 		ticks.push({
-			value: value,
-			major: major
+			value: value
```

Comment:
I prefer to push `{value: value, major: false}` to minimize the impact of the change.

Reply:
done